File: sbe/binance/spot_3_2/generated-python/spot_sbe/TPlusFilterLockEvent.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class UnlockData:
    """Repeating group item."""

    def __init__(self):
        self.unlock_time = None
        self.qty = None

class TPlusFilterLockEvent:
    """SBE message: TPlusFilterLockEvent."""

    TEMPLATE_ID = 608
    SCHEMA_ID = 3
    SCHEMA_VERSION = 2
    BLOCK_LENGTH = 11

    def __init__(self):
        self.event_time = None
        self.qty_exponent = None
        self.subscription_id = None
        self.unlock_data = []
        self.symbol = b''
        self.base_asset = b''

    def _decode_unlock_data_group(self, data: bytes, offset: int) -> Tuple[List[UnlockData], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<I', data, pos)[0]
        pos += 4

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = UnlockData()

            item.unlock_time = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.qty = struct.unpack_from('<q', data, pos)[0]
            pos += 8

            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def _decode_var_data(self, data: bytes, offset: int) -> Tuple[bytes, int]:
        """Decode variable-length binary data."""
        pos = offset
        length = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        value = data[pos:pos+length]
        pos += length
        return (value, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.event_time is not None:
            buffer.write(struct.pack('<q', self.event_time))
        if self.qty_exponent is not None:
            buffer.write(struct.pack('<b', self.qty_exponent))
        if self.subscription_id is not None:
            buffer.write(struct.pack('<H', self.subscription_id))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.event_time = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.qty_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.subscription_id = struct.unpack_from('<H', data, pos)[0]
        pos += 2

        # Skip to end of block for forward compatibility
        pos = 11

        self.unlock_data, pos = self._decode_unlock_data_group(data, pos)

        self.symbol, pos = self._decode_var_data(data, pos)
        self.base_asset, pos = self._decode_var_data(data, pos)

File: sbe/binance/spot_3_2/generated-php/spot_sbe/TPlusFilterLockEvent.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class UnlockData
{
    public int|float|null $unlockTime = null;
    public int|float|null $qty = null;
}

class TPlusFilterLockEvent
{
    public const TEMPLATE_ID = 608;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 2;
    public const BLOCK_LENGTH = 11;

    public int|float|array|null $eventTime = null;
    public int|float|array|null $qtyExponent = null;
    public int|float|array|null $subscriptionId = null;
    public array $unlockData = [];
    public string $symbol = '';
    public string $baseAsset = '';

    private function decodeUnlockDataGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new UnlockData();

            $item->unlockTime = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->qty = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;

            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    private function decodeVarData(string $data, int &$offset): string
    {
        $length = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;
        $value = substr($data, $offset, $length);
        $offset += $length;
        return $value;
    }

    public function encode(): string
    {
        $buffer = '';

        if ($this->eventTime !== null) {
            $buffer .= pack('q', $this->eventTime);
        }
        if ($this->qtyExponent !== null) {
            $buffer .= pack('c', $this->qtyExponent);
        }
        if ($this->subscriptionId !== null) {
            $buffer .= pack('v', $this->subscriptionId);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->eventTime = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->qtyExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->subscriptionId = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;

        // Skip to end of block for forward compatibility
        $offset = 11;

        $this->unlockData = $this->decodeUnlockDataGroup($data, $offset);

        $this->symbol = $this->decodeVarData($data, $offset);
        $this->baseAsset = $this->decodeVarData($data, $offset);
    }
}
